File: preprocess_data.m
clearvars; clc;
%%

% DATA_PATH = "C:\Users\tomma\Documents\Uni\PhD\data\Neurorobotics-data\2023trainval\";
DATA_PATH = "C:\Users\tomma\Documents\Uni\PhD\data\cBCI\all_subjects\";
DATA_PATH = "C:\Users\tomma\Documents\Uni\PhD\data\cBCI\sbj4_tr_rigorous\";
OUT_PATH = strcat(DATA_PATH,"preprocessed/");
FILT_ORDER = 2;
FC_l = 40;
FC_H = 2;
NUM_CHANNELS = 32;
LAP_PATH = "C:\Users\tomma\Documents\Uni\PhD\code\cBCI-MI-bhbf\config\lapmask_antneuro_32.mat";


WINSHIFT = 0.0625; %[s]
WINSIZE = 1; %[s]
Filenames = containers.Map (["train","val","test"],{ dir(strcat(DATA_PATH,"train/*.mat"));dir(strcat(DATA_PATH,"val/*.mat"));dir(strcat(DATA_PATH,"test/*.mat"));});
global EVENT
EVENT.START = 1;
EVENT.RIGHT_HAND = 769;
EVENT.LEFT_HAND = 770;
EVENT.BOTH_HANDS = 773;
EVENT.BOTH_FEET = 771;
EVENT.CONT_FEEDBACK = 781;
EVENT.REST = 783;
EVENT.FIXATION = 786;
EVENT.HIT = 897;
EVENT.MISS = 898;
EVENT.TIMEOUT = 899;

%%
if ~exist(OUT_PATH, 'dir')
   mkdir(OUT_PATH)
end

load(LAP_PATH,'lapmask');
for split = ["train" "val" "test"]
    printf("%s: reading from %s\n", split, strcat(DATA_PATH,split))
    eegs = cell(length(Filenames(split)),1);
    session_delims = [];
    event_pos = [];
    event_type = [];
    settings = eegc3_smr_newsettings();

    last_eeg_len = 0;
    dim_tot = 0;
    i=1;
    for file=Filenames(split)'
        disp(file.name)

        load(char(strcat(DATA_PATH,split,'/',file.name)),'s','h');
        t_eeg = s;
        t_header = h;
        if check_sanity(t_eeg,t_header)
            %% Read data, setup configuration
            events = t_header.EVENT;
    
            fs = t_header.SampleRate;
            winShift = fs*WINSHIFT;
            winSize = fs * WINSIZE;
            %% Preprocessing steps
            
            % Apply Laplacian spatial filter
            t_eeg = t_eeg(:,1:NUM_CHANNELS) * lapmask;
            % Apply bandpass filter
            for ch = 1:NUM_CHANNELS
                [b,a] = butter(FILT_ORDER,2*FC_l/fs,"low");
                t_eeg(:,ch) = filter(b,a,t_eeg(:,ch)); 
            
                [d,c] = butter(FILT_ORDER,2*FC_H/fs,"high");
                t_eeg(:,ch) = filter(d,c,t_eeg(:,ch)); 
            end

            eegs{i} = t_eeg;
            event_pos = [ event_pos; (t_header.EVENT.POS(:) + dim_tot)];
            event_type = [ event_type; (t_header.EVENT.TYP(:))];
            session_delims = [ session_delims; (last_eeg_len)+length(t_eeg)];
            dim_tot = dim_tot + length(t_eeg);
            i = i+1;
        else
            printf("Bad format, corrupt file or whatever, skipping..\n")
        end

    end
    %%
    EEG = nan(dim_tot, NUM_CHANNELS);
    pos = 0;
    for i=1:length(eegs)
        t_EEG = eegs{i};
        EEG(pos+(1:length(t_EEG)),:) = t_EEG;
        pos = pos+length(t_EEG);
    end

    printf("Total number of trials: %d\n", sum(event_type==EVENT.START));
    printf(" - Both feet: %d\n",sum(event_type==EVENT.BOTH_FEET));
    printf(" - Both hands: %d\n",sum(event_type==EVENT.BOTH_HANDS));
    printf(" - Rest: %d\n",sum(event_type==EVENT.REST));
    printf("%d Hits, %d misses\n",sum(event_type==EVENT.HIT),sum(event_type==EVENT.MISS));

    subj.settings = settings;
    subj.eeg = EEG;
    subj.triggers.pos = event_pos;
    subj.triggers.type = event_type;
    subj.session_delims = session_delims;
    save(strcat(OUT_PATH,"dataset_",split,".mat"),'subj','-v7.3');
end

%%

mask = event_type == EVENT.START | ...
       event_type == EVENT.BOTH_FEET |...
       event_type == EVENT.BOTH_HANDS |...
       event_type == EVENT.HIT |...
       event_type == EVENT.MISS;



function [sanity] = check_sanity(eeg,header)
    event_pos = header.EVENT.POS;
    event_type = header.EVENT.TYP;
    global EVENT
    if length(event_pos) ~= length(event_type)
        sanity = false;
        return
    end
    if sum(event_type==EVENT.START) ~= (sum(event_type==EVENT.HIT)+ ... 
                                        sum(event_type==EVENT.MISS)+ ...
                                        sum(event_type==EVENT.TIMEOUT))
        sanity = false;
        return
    end
    start_times = event_pos(event_type==EVENT.START);
    end_times = event_pos(event_type==EVENT.HIT | ...
                          event_type==EVENT.MISS |...
                          event_type==EVENT.TIMEOUT);
    diff = end_times -start_times;
    if any(diff<0)
        sanity=false;
        return
    end
    sanity=true;
end
%% Other
% Show EEG
% for i = 1:16
%     subplot(16,1,i)
%     plot(EEG(5000:length(EEG),i))
% end
% shg

%%Filtro 2-40hz
%%windows size 128
%%Sample size 32
%%overlap 0.5 s

File: preprocess_data.py
import glob
import os
import mne
import numpy as np
from neurorobotics_dl.events import Event
from neurorobotics_dl.utils import fix_mat
from scipy.io import loadmat,savemat
from scipy.signal import butter, lfilter

ALLOWED_FILE_FORMATS = ['mat','gdf']

"""
DATA_PATH = "/home/palatella/workspace/Neurorobotics-MI-data/d6"

OUT_PATH = os.path.join(DATA_PATH,"preprocessed")
LAP_PATH = "/home/palatella/workspace/BCI-MI-bhbf/config/laplacian16.mat"
# LAP_PATH = r"C:\Users\tomma\Documents\Uni\PhD\code\BCI-MI-bhbf\config\lapmask_antneuro_32.mat"
"""

DATA_PATH = r"C:\Users\tomma\Documents\Uni\PhD\data\d6"

OUT_PATH = os.path.join(DATA_PATH,"preprocessed")
LAP_PATH = r"C:\Users\tomma\Documents\Uni\PhD\code\BCI-MI-bhbf\config\laplacian16.mat"
# LAP_PATH = r"C:\Users\tomma\Documents\Uni\PhD\code\BCI-MI-bhbf\config\lapmask_antneuro_32.mat"
FILT_ORDER = 2
FC_H = 2
FC_L = 40
NUM_CHANNELS = 16
T_SHIFT = 0.0625
T_SIZE = 1

FILE_FORMATS = ["gdf","mat"]


def grab_files(root_dir,file_types=ALLOWED_FILE_FORMATS):
    files_grabbed = []
    for ftype in file_types:
        files_grabbed.extend(glob.glob(os.path.join(root_dir,f"*.{ftype}")))
    return files_grabbed

def main():
    for f in FILE_FORMATS:
        if f not in ALLOWED_FILE_FORMATS:
            raise Exception(f"Invalid file format {(f)} provided. Supported formats are", ','.join(ALLOWED_FILE_FORMATS))

    filenames = {"train": grab_files(os.path.join(DATA_PATH,"train")),
                "val": grab_files(os.path.join(DATA_PATH,"val")),
                "test": grab_files(os.path.join(DATA_PATH,"test"))}
    
    print()
    for k,v in filenames.items():
        print(f'{k}: found {len(v)} files')
    print()

    lap = loadmat(LAP_PATH)['lapmask']
    if not os.path.exists(OUT_PATH):
        os.makedirs(OUT_PATH)

    for split in ["train","val","test"]:
        if len(filenames[split])>0:
            print(f"{split}: reading from {os.path.join(DATA_PATH,split)}")
            eegs = [] 
            session_delims = []
            event_pos = []
            event_type = []
            last_eeg_len = 0
            dim_tot = 0
            i = 0

            for file in filenames[split]:
                print(file)
                file_format = file.split('.')[-1]
                if file_format == "gdf":
                    t_eeg,t_header = read_gdf(file)
                elif file_format == "mat":
                    t_eeg,t_header = read_mat(file)

                if check_sanity(t_eeg, t_header):
                    # Read data, setup configuration
                    events = t_header['EVENT']
                    fs = t_header['SampleRate']

                    # Preprocessing steps
                    # Apply Laplacian spatial filter
                    t_eeg = np.dot(t_eeg[:, :NUM_CHANNELS], lap)

                    # Apply bandpass filter
                    for ch in range(NUM_CHANNELS):
                        b, a = butter(FILT_ORDER, 2 * FC_L / fs, "low")
                        t_eeg[:, ch] = lfilter(b, a, t_eeg[:, ch])

                        d, c = butter(FILT_ORDER, 2 * FC_H / fs, "high")
                        t_eeg[:, ch] = lfilter(d, c, t_eeg[:, ch])
    
                    eegs.append(t_eeg)
                    event_pos.extend(events['POS'] + dim_tot)
                    event_type.extend(events['TYP'])
                    session_delims.append(last_eeg_len + len(t_eeg))
                    dim_tot += len(t_eeg)
                    i += 1
                else:
                    print("Bad format, corrupt file or whatever, skipping..")

            if dim_tot > 0:
                # EEG = np.full((dim_tot, NUM_CHANNELS), np.nan)
                # pos = 0
                # for i in range(len(eegs)):
                #     t_EEG = eegs[i]
                #     EEG[pos:(pos + len(t_EEG)), :] = t_EEG
                #     pos += len(t_EEG)
                EEG = np.concatenate(eegs)
                print(f"\nTotal number of trials: {np.sum(np.array(event_type) == Event.START)}")
                print(f" - Both feet: {np.sum(np.array(event_type) == Event.BOTH_FEET)}")
                print(f" - Both hands: {np.sum(np.array(event_type) == Event.BOTH_HANDS)}")
                print(f" - Rest: {np.sum(np.array(event_type) == Event.REST)}")
                print(f"{np.sum(np.array(event_type) == Event.HIT)} Hits, {np.sum(np.array(event_type) == Event.MISS)} misses")

                subj = {'eeg': EEG, 
                        'triggers': {'pos': event_pos, 'type': event_type},
                        'session_delims': session_delims}
                savemat(os.path.join(OUT_PATH, f"dataset_{split}.mat"), {'subj': subj})

def check_sanity(eeg,header):
    event_pos = header['EVENT']['POS']
    event_type = header['EVENT']['TYP']

    if sum(event_type == Event.CONT_FEEDBACK) == 0: # no trials found
        return False
    
    if len(event_pos) != len(event_type): # Some information has been lost during saving/loading v
        return False

    if sum(event_type == Event.CONT_FEEDBACK) != (sum(event_type == Event.HIT) +
                                          sum(event_type == Event.MISS) +
                                          sum(event_type == Event.TIMEOUT)): # Trial starts and trial ends are in different numbers
        return False

    start_times = event_pos[event_type == Event.CONT_FEEDBACK]
    end_times = event_pos[(event_type == Event.HIT) |
                         (event_type == Event.MISS) |
                         (event_type == Event.TIMEOUT)]
    diff = end_times - start_times
    if any(diff < 0): # Trial starts and trial ends are misaligned in time
        return False

    return True

def read_mat(spath):
        
    raw = loadmat(spath)
    t_eeg = raw['s']
    t_header = fix_mat(raw['h'])
    return t_eeg,t_header

def read_gdf(spath):
    raw=mne.io.read_raw_gdf(spath,verbose='error')
    events,names = mne.events_from_annotations(raw,verbose='error')
    names = {v:int(k) for k,v in names.items()}
    events_pos = events[:,0]
    events_typ = events[:,2]
    events_typ = [names[e] for  e in events_typ]
    t_eeg = raw.get_data().T

    t_header = {'SampleRate':raw.info['sfreq'],
                'EVENT':{'POS':np.array(events_pos),'TYP':np.array(events_typ)}
                }
    return t_eeg,t_header

if __name__=='__main__':
    main()